### ISSUE TITLE
Compiling Python packages with LPython

### ISSUE THREAD

**Issue by czgdp1807** [lcompilers/lpython#992]
Following is the plan to achieve the goal mentioned in issue title,

1. Write down small type annotated test packages in Python syntax. The nesting can vary from just 1 folder/package having an `__init__.py` file to 3-4 folders/subpackages one inside the other each level having an `__init__.py`. We can write packages for different sorting algorithms, graph algorithms and some practical stuff like that. We should try not to use workarounds but write code in a way that feels natural to us.
2. Try to compile those test packages starting from easy ones (with just 1 folder) to difficult ones. Do sprints to quickly make them work then send PRs each having a single fix.
3. Then proceed with small packages from PyPI (there must be many). Try to compile and fix LPython as described in step 2. 
4. Once all of the above is done, move on to advanced Python packages.

Alongside we should also try to interface with CPython code (#703).

@certik What do you say?

**Comment by Shaikh-Ubaid** [comment]
Interested!

**Comment by Shaikh-Ubaid** [comment]
Interested

**Comment by certik** [comment]
Yes, we should get LPython working with actual Python packages, that we write ourselves to ensure we stick to the subset that LPython supports. And the goal here is to fix up the inevitable bugs that we either know about or that we will hit. After most things work, it will allow people to upload packages to PyPi and to have dependencies, etc. We can use regular Python tooling to install dependencies and download/upload from PyPi. We have to improve LPython to be able to compile such a package with dependencies.

We probably also need to add a mode to compile the library and all dependencies to some kind of a "mod" file that stores the ASR in it, for quick compilation, so that when you modify a script, only that script has to be parsed and compiled to ASR, all other modules can be just deserialized. LFortran works this way, and it works great.

**Comment by czgdp1807** [comment]
I think we can re-use the serialization-deserialisation mechanism of LFortran and use it here to compile Python modules till ASR level. In fact, modules written by us right now compile (parsed + AST->ASR) only when we import them. However LFortran pre-compiles them till ASR and loads them if we import in our program. LPython should behave in a similar manner I think. Also, the question is how to integrate LPython with a build system (writing build files in simple words) so that LPython compiles all the intrinsic modules defined in it when its being built from source.

**Comment by certik** [comment]
Yes. I think CPython compiles every module into `.pyc`, so LPython would compile every module to `.lpy` (or even to `.pyc` to reuse the extension), and the compiled module would be just serialized ASR. Just like LFortran works with the `.mod` files.

**Comment by czgdp1807** [comment]
I see. Let me try implementing this idea then. Let's see how far we can get.

**Comment by czgdp1807** [comment]
In order to achieve first and second point in https://github.com/lcompilers/lpython/issues/992#issue-1344432422, I think we need to write `CMakeLists.txt` for the small test package. For now we can use `add_custom_command` to generate LLVM object files and then link those object files together.  So at the end for a single sub-package a static/shared binary will be generated. Does this make sense?

**Comment by certik** [comment]
For end users the main mode that we want to support is automatic compilation, just like CPython works. 

Is the CMake based mode just for testing that we can compile things by hand.

**Comment by certik** [comment]
We will probably support both modes:

* use cmake
* not use cmake, just do `lpython a.py` and if `a.py` imports a package `b`, the package `b` gets automatically compiled (and we can support some command line flag to compile just the package `b` into a list of `.pyc` files.)

I think most users would prefer the second approach. 

Or to rephrase it -- the cmake build system can be automatically generated by LPython, so why bother and not just compile things directly? We know all the information.

**Comment by czgdp1807** [comment]
I see. Makes sense. Second approach would be much easier. We will have to implement timestamp strategy to check whether the file has been changed or not. Do you know how to fetch the timestamp the last time a file was modified on disk a.k.a modification time? https://stackoverflow.com/a/40504396 seems to be a promising way to do get the modification time of a file but if you know of anything better then we can go for that as well.

**Comment by czgdp1807** [comment]
So I am planning to implement the strategy in my above comment. However the problem is that when you auto-compile and re-use the auto-compiled modules then symbol table IDs depend on the order in which you compile those modules. This would have consequences on `python run_tests.py -u` since it doesn't follow any particular order of compilation. Now say test `a.py` and `b.py` use a module `m.py` then if `a.py` gets compiled first then `m.py` will be different `b.py` and hence ASR output of `b.py` will not match. Something similar will happen if `b.py` gets compiled first. So how to deal with this situation? Because once a module is compiled then it changes the Symbol Table IDs of the rest of the program as well.

**Comment by certik** [comment]
This is dealt with in the Fortran desearilization by constructing a new symbol table and resolving it correctly. So everything should just work, as long as we reuse all the code, which I strongly recommend. :)

**Comment by certik** [comment]
Here is our goal:

* Imagine a project with 10 dependencies, each of which has 10 dependencies, so total of 200 packages have to be build
* The first time we build everything (say it takes 20s)
* Then when we modify one file in our project, the dependencies are not built again, only the file that changed is built and our project is rebuilt/linked (ideally this be immediate -- couple milliseconds; or as fast as possible)

**Comment by certik** [comment]
So what pieces do we need to get to the goal?

* We must be able to build 200 dependencies, and then reuse the build. 

What does that mean?

* We need to reuse as much as possible.
* Consequently, we have several options:
  * compile each file to .pyc (ASR serialization) or (even better): compile each package to one file that contains serialized ASR for the whole package
  * compile each file or (even better) each package into both serialized ASR and an object file(s)

The second option is more work for us, and we can tackle it later. We need something similar for LFortran as well, so we can work towards that as we go.

**Comment by certik** [comment]
Also what we need are nested modules. Say:
```
a (directory)
|
+- b.py
+- c.py
```
And we need to get the following working:
```python
from a.b import x
import a.b
from a import b
```
plus the relative import syntax in Python.

**Comment by certik** [comment]
Once the above works, we can think about how to build dependencies. I suggest we use `pyproject.toml` to specify the dependencies (Cargo style), see here for more background and links to other resources: https://stackoverflow.com/questions/62983756/what-is-pyproject-toml-file-for

**Comment by Thirumalai-Shaktivel** [comment]
@Shaikh-Ubaid are you working on this issue?

**Comment by certik** [comment]
Yes, @Shaikh-Ubaid is working on this issue, but if you are interested @Thirumalai-Shaktivel you can go ahead and also work on it, there is plenty of tasks to do.

**Comment by certik** [comment]
I just tested it, here is the latest status on this:

https://github.com/lcompilers/lpython/issues/1305

**Comment by Thirumalai-Shaktivel** [comment]
Related issues:
- #1317 
- #1280

**Comment by certik** [comment]
It's time to start writing LPython packages, upload to pypi, make them pip installable and see what breaks. Report issues that break and we'll fix it.

**Comment by certik** [comment]
Here are three ideas that we can do right away:
* Gauss-Legendre integration package, it gives back the weights and points, and you can use it to integrate, same interface as https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.roots_legendre.html
* A terminal package, that allows to use colors and that controls the terminal via escape codes
* A very simple minimization package, you give it a user function and it finds a minimum using some method.

**Comment by Smit-create** [comment]
I'll start with the first one.

**Comment by certik** [comment]
@Shaikh-Ubaid what are the instructions how to install some package using `pip`, and use (compile) it with LPython? Something along the lines of:
```console
conda create -n test1 python
conda activate test1
pip install XXX
lpython -I$CONDA_PREFIX/lib/python3.10/site-packages a.py
```
where `a.py` is something like:
```python
import XXX
XXX.something()
```
Let's get this working with some packages. Then let's iterate, perhaps something like this: `lpython --include-conda-env a.py`, or even a shorter option, perhaps just `lpython --conda a.py`.

**Comment by Shaikh-Ubaid** [comment]
>How many packages do we currently support? It looks like we can do:

There is also `nrp` (https://github.com/Shaikh-Ubaid/lpython_packages/tree/main/nrp_pkg)

```
Let's get a few more packages working.

Also let's iterate, perhaps something like this: lpython --include-conda-env a.py, or even a shorter option, perhaps just lpython --conda a.py. I would not allow environment variables by default yet, I want to get more experience using this first.
```
Sure.

**Comment by certik** [comment]
It looks like the `lpython` emulation works perfectly!

**Comment by certik** [comment]
I would version `lpython-emulation` exactly with the same version as the released LPython version. That way it is clear which one you have to use, depending on which features of LPython the script uses, as well as to avoid any backwards incompatibilities.

**Comment by certik** [comment]
I am upgrading this package: https://github.com/certik/mathfn/pull/5, I can see a few things:

* We should finish the lpython conda package to make it easier to install at the CI
* I want to depend on a few packages, maybe some terminal package, this mathfn package and then create a full end user application (binary)
* Polish the experience, make everything very easy

**Comment by Shaikh-Ubaid** [comment]
>I would version lpython-emulation exactly with the same version as the released LPython version.

We recently released LPython `0.14.0`. Do you mean we should version `lpython_emulation` as `0.14.0`? Do we update `lpython_emulation` only when `LPython` has a new release?

**Comment by certik** [comment]
That's what I would do. Do you see a problem with it?

**Comment by Shaikh-Ubaid** [comment]
> That's what I would do. Do you see a problem with it?

Got it. Sure, we can do that. It seems `lpython` does not actually use/depend-on `src/runtime/lpython/lpython.py`. `lpython` handles the types using `if-else` logical constructs in `AST->ASR`. I wonder if there is a possibility to distribute `src/runtime/lpython/lpython.py` (that is `lpython_emulation`) independently. For example, a use case can be a user who likes the `lpython_emulation` package and is wanting to use the package to add typing information to his `cpython` codes.

**Comment by certik** [comment]
LPython depends on `lpython.py` via integration_tests. We often upgrade a test, then we change `lpython.py` and AST->ASR. If we separate `lpython.py` from LPython, then we have to keep them in sync, which I think is more tedious than the current approach. Maybe later once LPython stabilizes we can do that.

**Comment by Shaikh-Ubaid** [comment]
>LPython depends on lpython.py via integration_tests. We often upgrade a test, then we change lpython.py and AST->ASR. If we separate lpython.py from LPython, then we have to keep them in sync, which I think is more tedious than the current approach. Maybe later once LPython stabilizes we can do that.

Got it, let's focus on delivering `lpython` first.
